app/controllers/locations_controller.rb

```diff
@@ -0,0 +1,43 @@
+class LocationsController < ApplicationController
+  def new
+    @location = Location.new
+    add_blank_ips!
+  end
+
+  def create
+    @location = Location.new(location_params_without_blank_ips)
+
+    if @location.save
+      Facades::Ips::AfterCreate.new.execute
+      redirect_to ips_path, notice: "#{@location.full_address} added"
+    else
+      add_blank_ips!
```

Comment:
Is this meant to be a bang method?
In Rails they are only used when they potentially raise an exception.

Reply:
This isn't my understanding of what `!` means on method names in Ruby, rather it's _"watch out, this could be dangerous!"_. 

In a ActiveRecord context it's often used for saying "this'll throw exceptions, watch out!". It often also points out it'll mutate something in-place (e.g. `sort` v `sort!`).

I've missed a key bit of idiomatic Ruby though - it's very rare to have `some_method!` defined without a safe version being available (`some_method`). I think I won't `!` here, even though modifying the params collection is a _little_ weird.

Reply:
Okay yeah fair enough, let's come back to this.
Needs an automated test wrapped around it.

Reply:
Woops this comment was meant for the location finding functionality.